Add a guide to writing a composition function in Python (#674)
crossplane/docs: content/knowledge-base/guides/write-a-composition-function-in-python.md

```

Your function's code lives in the `function` directory:

```shell {copy-lines=1}
ls function/
__version__.py  fn.py  main.py
```

The `function/fn.py` file is where you add the function's code. It's useful to
know about some other files in the template:

* `function/main.py` runs the function. You don't need to edit `main.py`.
* `Dockerfile` builds the function runtime. You don't need to edit `Dockerfile`.
* The `package` directory contains metadata used to build the function package.

{{<hint "tip">}}
<!-- vale gitlab.FutureTense = NO -->
<!--
This tip talks about future plans for Crossplane.
-->
In v1.14 of the Crossplane CLI `crossplane beta xpkg init` just clones a
template GitHub repository. In a future release the command will automate tasks
like replacing the template name with the new function's name. See Crossplane
issue [#4941](https://github.com/crossplane/crossplane/issues/4941) for details.
<!-- vale gitlab.FutureTense = YES -->
{{</hint>}}

Edit `package/crossplane.yaml` to change the package's name before you start
adding code. Name your package `function-xbuckets`.

> 

Author: I agree, but we're merging tooling to automate this any minute now (and I'm lazy). Once the tooling is merged I'll open another PR to cover using that instead of manually editing.

> Some functions accept a configuration input. You configure the input when you
> write a Composition that uses the function. The `package/input` directory
> defines the OpenAPI schema for the a function's input. 
> 
> The function in this guide doesn't accept an input. For this function you
> should delete the `package/input` directory. The
> [composition functions](https://docs.crossplane.io/latest/concepts/composition-functions)
> documentation explains more the input to a composition function.

Reviewer: ```suggestion
The `package/input` directory
defines the OpenAPI schema for the a function's input. 

The function in this guide doesn't accept an input. Delete the `package/input` 
directory.   
The
[composition functions](https://docs.crossplane.io/latest/concepts/composition-functions)
documentation explains more the input to a composition function.
```

I'd simplify the instructions and let the note below provide the context about inputs.

{{<hint "important">}}

> If you're writing a function that does use an input type, don't delete the
> `package/input` directory. Instead edit the file to be specific to your
> function.
> 
> The kind `Input` is a placeholder value. The API group
> `template.fn.crossplane.io` is, too. Change the kind and API group to something
> meaningful to your function. Edit the `openAPIV3Schema` to represent your
> function's input schema.

Reviewer: ```suggestion
Functions using inputs require the `package/input` directory. 
Edit the input YAML file to meet your function's requirements.

The `spec.names.kind: Input` and `spec.group: template.fn.crossplane.io` are 
placeholder values.  

Change the `kind` and `group` to something
meaningful to your function. Edit the `openAPIV3Schema` to represent your
function's input schema.
```

Simplify, make the language more direct

{{</hint>}}

## Edit the template to add the function's logic

You add your function's logic to the
{{<hover label="hello-world" line="1">}}RunFunction{{</hover>}}
method in `function/fn.py`. When you first open the file it contains a "hello
world" function.

```python {label="hello-world"}
async def RunFunction(self, req: fnv1beta1.RunFunctionRequest, _: grpc.aio.ServicerContext) -> fnv1beta1.RunFunctionResponse:
    log = self.log.bind(tag=req.meta.tag)
    log.info("Running function")

    rsp = response.to(req)

    example = ""
    if "example" in req.input:
        example = req.input["example"]

    # TODO: Add your function logic here!
    response.normal(rsp, f"I was run with input {example}!")
    log.info("I was run!", input=example)

    return rsp
```

> All Python composition functions have a `RunFunction` method. Crossplane passes
> everything the function needs to run in a
> {{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.
> The function tells Crossplane what resources it should compose by returning a
> {{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.

Reviewer: ```suggestion
All Python composition functions have a `RunFunction` method.  
Crossplane passes
everything the function needs to run in a
{{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.  
The function tells Crossplane what resources to compose by returning a
{{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.
```

Splitting across lines to improve readability.

I don't think your `{{<hover>}}` tags are matching what's intended.

{{<hint "tip">}}
Crossplane generates the `RunFunctionRequest` and `RunFunctionResponse` objects

> using [Protocol Buffers](https://protobuf.dev). You can find detailed schemas

Reviewer: ```suggestion
using [Protocol Buffers](https://protobuf.dev).  
You can find detailed schemas
```

Splitting lines for read/scanability.

for `RunFunctionRequest` and `RunFunctionResponse` in the
[Buf Schema Registry](https://buf.build/crossplane/crossplane/docs/main:apiextensions.fn.proto.v1beta1).
{{</hint>}}

Edit the `RunFunction` method to replace it with this code.

> ```python

Reviewer: ```suggestion
```python {hl_lines="7-28"}
```

Add static highlighting to make it obvious what's being changed and what's kept from the original template.

                {
                    "apiVersion": "s3.aws.upbound.io/v1beta1",
                    "kind": "Bucket",
                    "metadata": {
                        "annotations": {
                            "crossplane.io/external-name": name,
                        },
                    },
                    "spec": {
                        "forProvider": {
                            "region": region,
                        },
                    },
                }
            )

        # Log what the function did. This will only appear in the function's pod
        # logs. A function can use response.normal() and response.warning() to
        # emit Kubernetes events associated with the XR it's operating on.
        log.info("Added desired buckets", region=region, count=len(names))

        return rsp
```
{{</expand>}}

This code:

1. Gets the observed composite resource from the `RunFunctionRequest`.
1. Gets the region and bucket names from the observed composite resource.
1. Adds one desired S3 bucket for each bucket name.
1. Returns the desired S3 buckets in a `RunFunctionResponse`.

{{<hint "tip">}}

> Crossplane provides a
> [software development kit](https://github.com/crossplane/function-sdk-python)
> (SDK) for writing composition functions in Python. This function uses utilities
> from the SDK. Read the
> [documentation](https://crossplane.github.io/function-sdk-python) for the SDK.

Reviewer: ```suggestion
Crossplane provides a
[software development kit](https://github.com/crossplane/function-sdk-python)
(SDK) for writing composition functions in Python.  
This function uses utilities from the SDK. 

[Read the SDK documentation](https://crossplane.github.io/function-sdk-python) 
on the Crossplane GitHub repository.
```

Line breaks to improve readability. More descriptive link text.

{{</hint>}}

{{<hint "important">}}
The Python SDK automatically generates the `RunFunctionRequest` and
`RunFunctionResponse` Python objects from a
[Protocol Buffers](https://protobuf.dev) schema. You can see the schema in the
[Buf Schema Registry](https://buf.build/crossplane/crossplane/docs/main:apiextensions.fn.proto.v1beta1).

> The fields of the generated Python objects behave similarly to builtin Python
> types like dictionaries and lists. You should be aware that there are some

Reviewer: ```suggestion
The fields of the generated Python objects behave similar to builtin Python
types like dictionaries and lists. Be aware that there are some
```

Author: I think similarly is correct here based on a quick Googling.

Reviewer: "similarly" is an adverb describing verbs. In this context we're talking about a noun (builtin python type) so the adjective "similar" should be used.

differences.

Notably, you access the map of observed and desired resources like a dictionary
but you can't add a new desired resource by assigning to a map key. Instead,
access and mutate the map key as if it already exists.

Instead of adding a new resource like this:

```python
resource = {"apiVersion": "example.org/v1", "kind": "Composed", ...}
rsp.desired.resources["new-resource"] = fnv1beta1.Resource(resource=resource)
```

Pretend it already exists and mutate it, like this:

```python
resource = {"apiVersion": "example.org/v1", "kind": "Composed", ...}
rsp.desired.resources["new-resource"].resource.update(resource)
```

Refer to the Protocol Buffers
[Python Generated Code Guide](https://protobuf.dev/reference/python/python-generated/#fields)
for further details.
{{</hint>}}

## Test the function end-to-end

> You can test your function by adding unit tests, and by using the `crossplane
> beta render` command. It's a good idea to do both.

Reviewer: ```suggestion
Test your function with unit tests and with the `crossplane
beta render` command. 
```

                                        },
                                    }
                                )
                            ),
                        },
                    ),
                    context=structpb.Struct(),
                ),
            ),
        ]

        runner = fn.FunctionRunner()

        for case in cases:
            got = await runner.RunFunction(case.req, None)
            self.assertEqual(
                json_format.MessageToDict(got),
                json_format.MessageToDict(case.want),
                "-want, +got",
            )

if __name__ == "__main__":
    unittest.main()
```
{{</expand>}}

Run the unit tests using `hatch run`:

```shell {copy-lines="1"}
hatch run test:unit
.
----------------------------------------------------------------------
Ran 1 test in 0.003s

OK
```

{{<hint "tip">}}
[Hatch](https://hatch.pypa.io/) is a Python build tool. It builds Python
artifacts like wheels. It also manages virtual environments, similar
to `virtualenv` or `venv`. The `hatch run` command creates a virtual environment
and runs a command in that environment.

> You configure Hatch using `pyproject.toml`.

Reviewer: ```suggestion
```

Unless we're having them make changes I'd exclude this and leave it up to the Hatch docs to tell them where/what/how to customize.

{{</hint>}}

You can preview the output of a Composition that uses this function using
the Crossplane CLI. You don't need a Crossplane control plane to do this.

> Create a directory under `function-xbuckets` named `example`, and add the
> three files `xr.yaml`, `composition.yaml`, and `functions.yaml`.

Reviewer: ```suggestion
Create a directory under `function-xbuckets` named `example` and create a
Composite Resource, Composition and Function YAML files.  

Expand the following block to see example files. 
```

Tell them what they need to create then point to the examples.

{{<expand "The xr.yaml, composition.yaml and function.yaml files">}}

You can recreate the output below using by running `crossplane beta render` with
these files.

The `xr.yaml` file contains the composite resource to render:

```yaml
apiVersion: example.crossplane.io/v1
kind: XBuckets
metadata:
  name: example-buckets
spec:
  region: us-east-2
  names:
  - crossplane-functions-example-a
  - crossplane-functions-example-b
  - crossplane-functions-example-c
```

> 

Reviewer: ```suggestion
<br />

```

Fixing spacing due to weird markdown rendering.

The `composition.yaml` file contains the Composition to use to render the
composite resource:

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: create-buckets
spec:
  compositeTypeRef:
    apiVersion: example.crossplane.io/v1
    kind: XBuckets
  mode: Pipeline
  pipeline:
  - step: create-buckets
    functionRef:
      name: function-xbuckets
```

> 

Reviewer: ```suggestion
<br />

```

Fixing spacing due to weird markdown rendering.

The `functions.yaml` file contains the Functions the Composition references in
its pipeline steps:

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: function-xbuckets
  annotations:
    render.crossplane.io/runtime: Development
spec:
  # The CLI ignores this package when using the Development runtime.
  # You can set it to any value.
  package: xpkg.upbound.io/negz/function-xbuckets:v0.1.0
```
{{</expand>}}

> Note that the Function in `functions.yaml` uses the

Reviewer: ```suggestion
The Function in `functions.yaml` uses the
```

{{<hover label="development" line="6">}}Development{{</hover>}}
runtime. This tells `crossplane beta render` that your function is running
locally. It connects to your locally running function instead of using Docker to
pull and run the function.

```yaml {label="development"}
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: function-xbuckets
  annotations:
    render.crossplane.io/runtime: Development
```

> Use `hatch run development` to run your function locally. This tells the
> function to run without encryption or authentication. You should only use it
> during testing and development.

Reviewer: ```suggestion
Use {{<hover line="1" label="run">}}hatch run development{{</hover>}} 
to run your function locally.  
The `development` argument runs the
function without encryption or authentication. 

{{<hint "warning" >}}
Only use `development` during testing and development.
{{< /hint >}}
```

Use `hover` to call draw attention to the command because of the number of nearby code boxes.

Pulls the security warning into a dedicated `warning` box.

world" function.

```python {label="hello-world"}
class FunctionRunner(grpcv1beta1.FunctionRunnerService):
    def __init__(self):
        self.log = logging.get_logger()

    async def RunFunction(
        self, req: fnv1beta1.RunFunctionRequest, _: grpc.aio.ServicerContext
    ) -> fnv1beta1.RunFunctionResponse:
        log = self.log.bind(tag=req.meta.tag)
        log.info("Running function")

        rsp = response.to(req)

        example = ""
        if "example" in req.input:
            example = req.input["example"]

        # TODO: Add your function logic here!
        response.normal(rsp, f"I was run with input {example}!")
        log.info("I was run!", input=example)

        return rsp
```

All Python composition functions have a `RunFunction` method. Crossplane passes
everything the function needs to run in a
{{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.
The function tells Crossplane what resources it should compose by returning a
{{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.

{{<hint "tip">}}
Crossplane generates the `RunFunctionRequest` and `RunFunctionResponse` objects

> using [Protocol Buffers](https://protobuf.dev). You can find detailed schemas

Author: Vale was complaining about `protobuf`, despite it being part of a link URL. I tried putting the link on a line by itself without success. In the end I added it as allowed jargon, since I could imagine us wanting to say "protobuf" rather than "Protocol Buffers" outside of links too.

Author: Ah, I did actually use `protobuf` once outside a link down on line 357. I'm guessing it was complaining about that, but incorrectly telling me the error was coming from this line for some reason.

Reviewer: I've seen that behavior before and it's a bug I haven't been able to chase down. I haven't been able to figure out what Vale is keying on to put the error in the wrong place.
